src/util/server.ml

```diff
@@ -0,0 +1,175 @@
+  type response
+
+  val params_of_yojson: Yojson.Safe.t -> (params, string) result
+  val response_to_yojson: response -> Yojson.Safe.t
+
+  val process: params -> t -> response
+end
+
+module Registry = struct
+  type t = (string, (module Request)) Hashtbl.t
+  let make () : t = Hashtbl.create 32
+  let register (reg: t) (module R : Request) = Hashtbl.add reg R.name (module R)
+end
+
+let registry = Registry.make ()
+
+let handle_exn id exn =
+  Response.Error.(make Code.InternalError (Printexc.to_string exn) () |> Response.error id)
+
+module ParamParser (R : Request) = struct
+  let parse params =
+    let maybe_params =
+      params
+      |> Option.map_default Message.Structured.to_json `Null
+      |> R.params_of_yojson
+    in
+    match maybe_params with
+    | Ok params -> Ok params
+    | Error err ->
+      (* This is a hack to handle cases where R.params is a primitive type like int or string. *)
+      match params with
+      | Some `List [param] -> R.params_of_yojson param |> Result.map_error (fun _ -> err)
+      | _ -> Error err
+end
+
+let handle_request (serv: t) (message: Message.either) (id: Id.t) =
+  let req = Hashtbl.find_option registry message.method_ in
+  let response = match req with
+    | Some (module R) ->
+      let module Parser = ParamParser (R) in (
+        match Parser.parse message.params with
+        | Ok params -> (
+            try
+              R.process params serv
+              |> R.response_to_yojson
+              |> Response.ok id
+            with Failure (code, msg) -> Response.Error.(make code msg () |> Response.error id))
+        | Error s -> Response.Error.(make Code.InvalidParams s () |> Response.error id))
+    | _ -> Response.Error.(make Code.MethodNotFound message.method_ () |> Response.error id)
+  in
+  Response.yojson_of_t response |> Yojson.Safe.to_string |> IO.write_line serv.output;
+  IO.flush serv.output
+
+let serve serv =
+  serv.input
+  |> IO.to_input_channel
+  |> Yojson.Safe.linestream_from_channel
+  |> Stream.iter (fun line ->
+      match line with
+      | `Json json -> (
+          try
+            let message = Message.either_of_yojson json in
+            match message.id with
+            | Some id -> handle_request serv message id
+            | _ -> () (* We just ignore notifications for now. *)
+          with Json.Of_json (s, _) -> prerr_endline s)
+      | `Exn exn -> prerr_endline (Printexc.to_string exn))
+
+let make ?(input=stdin) ?(output=stdout) file do_analyze : t = { file; do_analyze; input; output }
+
+let bind () =
+  let server = GobConfig.get_string "server" in
+  if server = "-" then None, None else (
+    if Sys.file_exists server then
+      Sys.remove server;
+    let socket = Unix.socket PF_UNIX SOCK_STREAM 0 in
+    Unix.bind socket (ADDR_UNIX server);
+    Unix.listen socket 1;
+    let conn, _ = Unix.accept socket in
+    Unix.close socket;
+    Sys.remove server;
+    Some (Unix.input_of_descr conn), Some (Unix.output_of_descr conn))
```

Comment:
The way I understand it, this currently opens the UNIX socket once to accept one client and none any further. It might be nice if the UNIX socket remained and accepted another connection after the first one closes. So no parallel handling of multiple connections, but just sequentially. This could be useful if you want to reconnect a client to the server without having to kill the server.

Reply:
Yeah, that is what it does. My main goal was to simplify the process clean-up in case the parent HTTP server dies. But I guess it isn't complicated to make the parent process terminate its children before crashing.

Reply:
Depending how the parent process dies (e.g. if it's killed or crashes), then it might not have the opportunity to terminate its children anymore.
I think for the time being we can leave it as-is until there's actual need to reconnect to the server.